sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,855 @@
+	}
+
+	/* FIXME: Change card detection pin polarity to make
+	 * the embedded module work on dataro. CD_POL bit is
+	 * marked as "Only for debug". This needs changing.
+	 * i.MX 6ULL Reference manual rev.1, page 4072. */
+	*(sdio_common.base + reg_vend_spec) |= (1 << 5);
+
+	*(sdio_common.base + reg_int_status_en) = SDHC_INTERRUPT_DEFAULTS;
+
+	for (i = 0; i < 5; i++) {
+		rslt = sdio_cardInit();
+		if (rslt == 0) {
+			break;
+		}
+		usleep(1000);
+	}
+	if (rslt < 0) {
+		printf("libsdio: failed to initialize SDIO card\n");
+		sdio_free();
+		return -1;
+	}
+
+	sdio_common.sdioInitialized = 1;
+	return 0;
+}
+
+
+int sdio_config(uint32_t freq, uint16_t blocksz)
+{
+	uint8_t i;
+	uint32_t val;
+
+	if (freq != 25000000 && freq != 50000000) {
+		printf("libsdio: unsupported clock frequency (%uHz)\n", freq);
+		return -1;
+	}
+
+	sdio_common.blocksz = blocksz;
+
+	*(sdio_common.base + reg_prot_ctrl) |= (1 << 1);  /* set 4-bit mode */
+	*(sdio_common.base + reg_vend_spec) &= ~(1 << 8); /* FRC_SDCLK_ON=0 */
+	/* "Before changing clock divisor [...] Host Driver
+	 * should make sure the SDSTB bit is high"
+	 * i.MX 6ULL Reference Manual Rev.1, page 4028 */
+	for (i = 0; i < SDHC_RETRIES; i++) {
+		val = *(sdio_common.base + reg_pres_state) & (1 << 3);
+		if (val != 0) {
+			break;
+		}
+		usleep(10);
+	}
+	if (val == 0) {
+		return -1;
+	}
+
+	if (freq == 50000000) {
+		*(sdio_common.base + reg_sys_ctrl) = SDHC_SYSCTL_CLK_DIV_4 | SDHC_SYSCTL_RESERVED;
+	}
+	else {
+		*(sdio_common.base + reg_sys_ctrl) = SDHC_SYSCTL_CLK_DIV_8 | SDHC_SYSCTL_RESERVED;
+	}
+
+	/* Send 80 initialization clock cycles
+	 * to the card, and wait until it finishes.
+	 * Should be done in ~3us on 25MHz clock */
+	*(sdio_common.base + reg_sys_ctrl) |= (1 << 27);
+	for (i = 0; i < SDHC_RETRIES; i++) {
+		val = *(sdio_common.base + reg_sys_ctrl) & (1 << 27);
+		if (val == 0) {
+			break;
+		}
+		usleep(10);
+	}
+	if (val != 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+void sdio_free(void)
+{
+	sdhc_reset(SDHC_SYSCTL_RESET_ALL);
+
+	/*
+	 * FIXME: the following resources:
+	 * isrHandle, eventLock, eventCond
+	 * are not destroyed since they are used
+	 * by running event service thread.
+	 * The same applies for hardware register
+	 * base address which remains mapped.
+	 * this is due to supposed problems with
+	 * threadJoin function which would allow
+	 * to stop the thread properly.
+	 */
+
+	sdio_common.sdioInitialized = 0;
+	sdio_common.blocksz = 0;
+
+	if (sdio_common.cmdLock >= 0) {
+		resourceDestroy(sdio_common.cmdLock);
+		sdio_common.cmdLock = (handle_t)NULL;
+	}
+	if (sdio_common.dmaptr != NULL) {
+		munmap(sdio_common.dmaptr, DMA_BUFFER_SIZE);
+		sdio_common.dmaptr = NULL;
+		sdio_common.dmaphys = (addr_t)NULL;
+	}
+
+	*(sdio_common.base + reg_int_status_en) = 0;
+	*(sdio_common.base + reg_int_signal_en) = 0;
+
+	memset(&sdio_common.eventHandlers, 0, sizeof(sdio_common.eventHandlers));
+	memset(&sdio_common.eventHandlerArgs, 0, sizeof(sdio_common.eventHandlerArgs));
```

Comment:
At the moment it is not thread-safe, an event can processed at the moment, so better do it under an `eventLock`.

Reply:
Appropriate thread security measures implemented.